kotlinx-coroutines-core/common/src/flow/SharedFlow.kt

```diff
@@ -0,0 +1,654 @@
+    private fun cleanupTailLocked() {
+        // If we have synchronous case, then keep one emitter queued
+        if (bufferCapacity == 0 && queueSize <= 1) return // return, don't clear it
+        val buffer = buffer!!
+        while (queueSize > 0 && buffer.getBufferAt(head + totalSize - 1) === NO_VALUE) {
+            queueSize--
+            buffer.setBufferAt(head + totalSize, null)
+        }
+    }
+
+    // returns NO_VALUE if cannot take value without suspension
+    private fun tryTakeValue(slot: SharedFlowSlot): Any? {
+        var resumeList: List<Continuation<Unit>>? = null
+        val value = synchronized(this) {
+            val index = tryPeekLocked(slot)
+            if (index < 0) {
+                NO_VALUE
+            } else {
+                val oldIndex = slot.index
+                val newValue = getPeekedValueLockedAt(index)
+                slot.index = index + 1 // points to the next index after peeked one
+                resumeList = updateCollectorIndexLocked(oldIndex)
+                newValue
+            }
+        }
+        resumeList?.forEach { it.resume(Unit) }
+        return value
+    }
+
+    // returns -1 if cannot peek value without suspension
+    private fun tryPeekLocked(slot: SharedFlowSlot): Long {
+        // return buffered value if possible
+        val index = slot.index
+        if (index < bufferEndIndex) return index
+        if (bufferCapacity > 0) return -1L // if there's a buffer, never try to rendezvous with emitters
+        // Synchronous shared flow (bufferCapacity == 0) tries to rendezvous
+        if (index > head) return -1L // ... but only with the first emitter (never look forward)
+        if (queueSize == 0) return -1L // nothing there to rendezvous with
+        return index // rendezvous with the first emitter
+    }
+
+    private fun getPeekedValueLockedAt(index: Long): Any? =
+        when (val item = buffer!!.getBufferAt(index)) {
+            is Emitter -> item.value
+            else -> item
+        }
+
+    private suspend fun awaitValue(slot: SharedFlowSlot): Unit = suspendCancellableCoroutine { cont ->
+        synchronized(this) lock@{
+            val index = tryPeekLocked(slot) // recheck under this lock
+            if (index < 0) {
+                slot.cont = cont // Ok -- suspending
+            } else {
+                cont.resume(Unit) // has value, no need to suspend
+                return@lock
+            }
+            slot.cont = cont // suspend, waiting
+        }
+    }
+
+    private fun findSlotsToResumeLocked(): List<Continuation<Unit>>? {
+        var result: ArrayList<Continuation<Unit>>? = null
+        forEachSlotLocked loop@{ slot ->
+            val cont = slot.cont ?: return@loop // only waiting slots
+            if (tryPeekLocked(slot) < 0) return@loop // only slots that can peek a value
+            val a = result ?: ArrayList<Continuation<Unit>>(2).also { result = it }
```

Comment:
Why 2?

Reply:
Optimized for common cases of very few subscribers. I expect that for most shared flows there will be, in fact, only a single subscriber. Now it is rewritten to arrays, still allocates size-2 to start with